### ISSUE TITLE
Why the features extracted from mxnet-cpp Windows API different from that extracted from mxnet Linux?

### ISSUE THREAD

**Issue by lzhaozi** [apache/incubator-mxnet#4170]
For bugs or installation issues, please provide the following information.
The more information you provide, the more likely people will be able to help you.

## Environment info
Operating System:
Windows 10 professional
Compiler:
Microsoft visual studio 2013
Package used (Python/R/Scala/Julia):

MXNet version:
mxnet-cpp
Or if installed from source:

MXNet commit hash (`git rev-parse HEAD`):

If you are using python package, please provide

Python version and distribution:

If you are using R package, please provide

R `sessionInfo()`:

## Error Message:
Please paste the full error message, including stack trace.

## Minimum reproducible example
if you are using your own code, please provide a short script that reproduces the error.

## Steps to reproduce
or if you are running standard examples, please provide the commands you have run that lead to the error.

MXNet Windows code:
#include <iostream>
#include <vector>
#include <string>
#include "mxnet-cpp/MxNetCpp.h"
#include <fstream>
#include <map>

using namespace std;
using namespace mxnet::cpp;

#include <opencv2/opencv.hpp>
using namespace cv;

const int K = 4; //topK

/*The global context, change them if necessary*/
// Context global_ctx(kGPU, 0);
Context global_ctx(kCPU, 0);

class FeatureExtractor {
private:
	/*the mean image, get from the pretrained model*/
	NDArray mean_img;
	/*the following two maps store all the paramters need by the model*/
	map<string, NDArray> args_map;
	map<string, NDArray> aux_map;
	Symbol net;
	int channel;
	Executor *executor;
	/*Get the feature layer we want to extract*/
	void GetFeatureSymbol(string json_net) {
		/*
		* use the following to check all the layers' names:
		* */
		
		//net=Symbol::Load(json_net).GetInternals();
		//for(const auto & layer_name:net.ListOutputs()){
		//cout<<layer_name<<endl;
		//}
		
		net = Symbol::Load(json_net).GetInternals()["global_pool_output"];
	}
	/*Fill the trained paramters into the model, a.k.a. net, executor*/
	void LoadParameters(string params) {
		map<string, NDArray> paramters;
		NDArray::Load(params, 0, &paramters);
		for (const auto &k : paramters) {
			if (k.first.substr(0, 4) == "aux:") {
				auto name = k.first.substr(4, k.first.size() - 4);
				aux_map[name] = k.second.Copy(global_ctx);
			}
			if (k.first.substr(0, 4) == "arg:") {
				auto name = k.first.substr(4, k.first.size() - 4);
				args_map[name] = k.second.Copy(global_ctx);
			}
		}
		/*WaitAll is need when we copy data between GPU and the main memory*/
		NDArray::WaitAll();
	}
	void GetMeanImg(string mean) {
		mean_img = NDArray(Shape(1, channel, 299, 299), global_ctx, false);
		mean_img.SyncCopyFromCPU(
			NDArray::LoadToMap(mean)["mean_img"].GetData(),
			1 * channel * 299 * 299);
		NDArray::WaitAll();
	}

public:
	FeatureExtractor() {
		/*prepare the model, fill the pretrained parameters, get the mean image*/
		channel = 3;
		GetFeatureSymbol("./model/Inception_v3c.json");
		LoadParameters("./model/Inception_v3c-20.params");
		GetMeanImg("./model/meanc.bin");
	}

	FeatureExtractor(int c, string json_net, string params, string mean) {
		channel = c;
		GetFeatureSymbol(json_net);
		LoadParameters(params);
		GetMeanImg(mean);
	}

	vector<float> Extract(NDArray data) {
		/*Normalize the pictures*/
		data -= mean_img;
		//data.Slice(0, 1) -= mean_img;
		//data.Slice(1, 2) -= mean_img;
		args_map["data"] = data;
		/*bind the excutor*/
		executor = net.SimpleBind(global_ctx, args_map, map<string, NDArray>(),
			map<string, OpReqType>(), aux_map);
		executor->Forward(false);
		/*print out the features*/
		auto array = executor->outputs[0].Copy(Context(kCPU, 0));
		NDArray::WaitAll();
		
		vector<float> feat;
		cout << array.Size() << endl; // 2048
		
		for (int i = 0; i < array.Size(); ++i) {
			feat.push_back(array.At(0, i));
			//cout << array.At(0, i) << ",";
		}
		return feat;
	}
};
/*read images and store them the NDArray format that MXNet.cpp can handle*/
NDArray Mat2Array(string filename, int channel) {
	string file_name_list[] = { filename };

	std::vector<float> array;
	for (auto &t : file_name_list) {
		cv::Mat mat = cv::imread(t);
		/*resize pictures to (224, 224) according to the pretrained model*/
		cv::resize(mat, mat, cv::Size(299, 299));
		for (int c = 0; c < channel; ++c) {
			for (int i = 0; i < 299; ++i) {
				for (int j = 0; j < 299; ++j) {
					array.push_back(static_cast<float>(mat.data[(i * 299 + j) * channel + c]));
				}
			}
		}
	}
	NDArray ret(Shape(1, channel, 299, 299), global_ctx, false);
	ret.SyncCopyFromCPU(array.data(), 1 * channel * 299 * 299);
	NDArray::WaitAll();
	return ret;
}

int main()
{
	string filename("./testc.jpg");
	auto data = Mat2Array(filename, 3);

	FeatureExtractor fe(3, "inception-v3c.json", "inception-v3c-20.params", "meanc.bin");
	vector<float> feat(fe.Extract(data));
	return 0;
}

mxnet Linux code:
import mxnet as mx
import logging
import numpy as np
import scipy.io as sio
import time

logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

batch_size = 1

test_dataiter = mx.io.ImageRecordIter(
        path_imgrec="testc.rec",
        mean_img="rgbd/meanc.bin",
        rand_crop=False,
        rand_mirror=False,
        data_shape=(3,299,299),
        batch_size=batch_size,
        round_batch=False,
        preprocess_threads=1)

prefix = "models/inception-v3/inception-v3c"
num_epoch = 20

model = mx.model.FeedForward.load(prefix, num_epoch, ctx=mx.gpu())

internals = model.symbol.get_internals()

fea_symbol = internals["global_pool_output"]

feature_extractor = mx.model.FeedForward(ctx=mx.gpu(), symbol=fea_symbol,
                                         arg_params=model.arg_params,
                                         aux_params=model.aux_params,
                                         allow_extra_params=True)
global_pooling_feature = feature_extractor.predict(test_dataiter)

Unfortunately, the feat has the same 2048 dim as global_pooling_feature but has completely different content as it, making the performance much worse in Windows than on Linux. Does the mxnet Windows api has bugs?

**Comment by piiswrong** [comment]
did you use prebuilt binaries? Binaries doesn't work for some versions of windows due to cuda dll compatibility issues.

See if it works with ctx=mx.cpu(0)

**Comment by kwanking** [comment]
I also failed to extract feature by pertained model. 
according to some documents in internet, it is bgr in opencv and rgb in mxnet.

I have a question, would you like tell me how do get meanc.bin?

**Comment by yajiedesign** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
